diff --git a/tools/9pfsd/Makefile b/tools/9pfsd/Makefile
index b5f9703131..99a8021fd4 100644
--- a/tools/9pfsd/Makefile
+++ b/tools/9pfsd/Makefile
@@ -14,7 +14,7 @@ XEN-9PFSD_OBJS = xen-9pfsd.o io.o
 $(XEN-9PFSD_OBJS): CFLAGS += $(CFLAGS_libxenstore)
 $(XEN-9PFSD_OBJS): CFLAGS += $(CFLAGS_libxenevtchn)
 $(XEN-9PFSD_OBJS): CFLAGS += $(CFLAGS_libxengnttab)
-xen-9pfsd: LDLIBS += $(call xenlibs-ldlibs,store evtchn gnttab)
+xen-9pfsd: LDLIBS += $(call xenlibs-ldlibs,store evtchn gnttab toolcore toollog)
 
 .PHONY: all
 all: $(TARGETS)
