diff --git a/hw/9pfs/xen-9p-backend.c b/hw/9pfs/xen-9p-backend.c
index 79359d911a..7152c84bca 100644
--- a/hw/9pfs/xen-9p-backend.c
+++ b/hw/9pfs/xen-9p-backend.c
@@ -490,7 +490,7 @@ static int xen_9pfs_connect(struct XenLegacyDevice *xendev)
     qemu_opt_set(fsdev, "fsdriver", "local", NULL);
     qemu_opt_set(fsdev, "path", xen_9pdev->path, NULL);
     qemu_opt_set(fsdev, "security_model", xen_9pdev->security_model, NULL);
-    qemu_opts_set_id(fsdev, s->fsconf.fsdev_id);
+    qemu_opts_set_id(fsdev, strdup(s->fsconf.fsdev_id));
     qemu_fsdev_add(fsdev, &err);
     if (err) {
         error_report_err(err);
